README.md
![syncfusion_flutter_slider_banner](https://cdn.syncfusion.com/content/images/FTControl/Flutter/sliders/slider-banner.png)

# Syncfusion Flutter Sliders

Syncfusion Flutter Sliders library is written natively in Dart for creating highly interactive and UI-rich slider controls for filtering purposes in Flutter applications.

## Overview

This library is used to create three different types of sliders, namely slider, range slider, and range selector. All these sliders have a rich set of features such as support for both numeric and date values, tooltip, labels, and ticks. The range selector latter accepts any kind of child including [Charts](https://pub.dev/packages/syncfusion_flutter_charts).

**Disclaimer:** This is a commercial package. To use this package, you need to have either a Syncfusion Commercial License or Syncfusion Community license. For more details, please check the [LICENSE](https://github.com/syncfusion/flutter-examples/blob/master/LICENSE) file.

**Note:** Our packages are now compatible with Flutter for Web. However, this will be in Beta until Flutter for Web becomes stable.

## Table of contents
- [Slider features](#slider-features)
- [Range slider features](#range-slider-features)
- [Range selector features](#range-selector-features)
- [Get the demo application](#get-the-demo-application)
- [Useful links](#useful-links)
- [Installation](#installation)
- [Slider getting started](#slider-getting-started)
  - [Add slider to the widget tree](#add-slider-to-the-widget-tree)
  - [Add slider elements](#add-slider-elements)
- [Range slider getting started](#range-slider-getting-started)
  - [Add range slider to the widget tree](#add-range-slider-to-the-widget-tree)
  - [Add range slider elements](#add-range-slider-elements)
- [Range selector getting started](#range-selector-getting-started)
  - [Add range selector to the widget tree](#add-range-selector-to-the-widget-tree)
  - [Add range selector elements](#add-range-selector-elements)
- [Support and Feedback](#support-and-feedback)
- [About Syncfusion](#about-syncfusion)

## Slider features

* **Numeric and date support** - Select numeric and date value. For date, support is provided up to the 'Seconds' interval.

* **Labels** - Render labels for the date and numeric ranges with an option to customize formats based on the requirement.
![slider labels](https://cdn.syncfusion.com/content/images/Flutter/pub_images/slider_images/slider_labels.png)

* **Ticks and divisors** - Show ticks and divisors based on the interval. Also, enable minor ticks between the ticks to indicate the selected values. These options show the selected range in a more intuitive way for the end users.
![slider ticks](https://cdn.syncfusion.com/content/images/Flutter/pub_images/slider_images/slider_ticks.png)

* **Tooltip** - Render tooltip to show the selected value clearly. It is also possible to customize the format of the text shown in the tooltip.
![slider tooltip](https://cdn.syncfusion.com/content/images/Flutter/pub_images/slider_images/slider_tooltip.png)

* **Thumb icon support** - Accepts custom widgets like icon or text inside the thumb.
![slider thumb icon](https://cdn.syncfusion.com/content/images/Flutter/pub_images/slider_images/slider_icon.png)

* **Discrete selection** - Provides an option for selecting only discrete numeric and date values.

* **Highly customizable** - In addition to the built-in rich set of features, fully customize the control easily using the wide range of provided options.
![slider customization](https://cdn.syncfusion.com/content/images/Flutter/pub_images/slider_images/slider_customization.png)

## Range slider features

Range slider supports all the above-mentioned features of the slider in addition to:

* **Interval selection** - Allows users to select a particular interval by tapping or clicking in it. Both thumbs will be moved to the current interval with animation.

## Range selector features

Range selector supports all the above-mentioned features of the range slider in addition to:

* **Child support** - Add a child of any type inside the range selector. It is also possible to add Charts. With the built-in integration, range selector is smart enough to handle features like segment selection or zooming of a chart based on the selected range in the range selector. Similar to the range slider, it also supports both numeric and date values.

* **Deferred update** - Provides an option to defer range updates, allowing you to control when dependent components are updated while thumbs are being dragged continuously.
![range selector](https://cdn.syncfusion.com/content/images/FTControl/Flutter/sliders/range-selector.gif)

## Get the demo application

Explore the full capability of our Flutter widgets on your device by installing our sample browser application from the following app stores. View sample codes in GitHub.

<p align="center">
  <a href="https://play.google.com/store/apps/details?id=com.syncfusion.flutter.examples"><img src="https://cdn.syncfusion.com/content/images/FTControl/google-play.png"/></a>
  <a href="https://apps.apple.com/us/app/syncfusion-flutter-ui-widgets/id1475231341"><img src="https://cdn.syncfusion.com/content/images/FTControl/apple-button.png"/></a>
</p>
<p align="center">
  <a href="https://github.com/syncfusion/flutter-examples"><img src="https://cdn.syncfusion.com/content/images/FTControl/GitHub.png"/></a>
  <a href="https://flutter.syncfusion.com"><img src="https://cdn.syncfusion.com/content/images/FTControl/web_sample_browser.png"/></a>
</p>

## Useful links
Take a look at the following to learn more about Syncfusion Flutter sliders:

* [Syncfusion Flutter Sliders product page](https://www.syncfusion.com/flutter-widgets)
* [User guide documentation for Slider](https://help.syncfusion.com/flutter/slider)
* [User guide documentation for Range Slider](https://help.syncfusion.com/flutter/range-slider)
* [User guide documentation for Range Selector](https://help.syncfusion.com/flutter/range-selector)
* [Knowledge base](https://www.syncfusion.com/kb)

## Installation

Install the latest version from [pub](https://pub.dev/packages/syncfusion_flutter_sliders#-installing-tab-).

## Slider getting started

Import the following package.

```dart
import 'package:syncfusion_flutter_sliders/sliders.dart';
```

### Add slider to the widget tree

Add the slider widget as a child of any widget. Here, the slider widget is added as a child of the center widget.

```dart
@override
Widget build(BuildContext context) {
    return MaterialApp(
        home: Scaffold(
            body: Center(
                child: SfSlider(),
              ),
          ),
      );
}
```

### Add slider elements

Add the slider elements such as ticks, labels, and tooltip to show the current position of the slider thumb.

```dart
double _value = 40.0;

@override
Widget build(BuildContext context) {
  return Scaffold(
     appBar: AppBar(
       title: const Text('Syncfusion Flutter Slider'),
     ),
     body: SfSlider(
       min: 0.0,
       max: 100.0,
       value: _value,
       interval: 20,
       showTicks: true,
       showLabels: true,
       enableTooltip: true,
       minorTicksPerInterval: 1,
       onChanged: (dynamic value){
         setState(() {
           _value = value;
         });
       },
     ),
   );
}
```

The following screenshot illustrates the result of the above code sample.

![simple slider](https://cdn.syncfusion.com/content/images/Flutter/pub_images/slider_images/Slider_Tooltip.png)

## Range slider getting started

Import the following package.

```dart
import 'package:syncfusion_flutter_sliders/sliders.dart';
```

### Add range slider to the widget tree

Add the range slider widget as a child of any widget. Here, the range slider widget is added as a child of the center widget.

```dart
@override
Widget build(BuildContext context) {
    return MaterialApp(
        home: Scaffold(
            body: Center(
                child: SfRangeSlider(),
              ),
          ),
      );
}
```

### Add range slider elements

Add the range slider elements such as ticks, labels, and tooltips to show the current position of the range slider thumb.

```dart
SfRangeValues _values = SfRangeValues(40.0, 80.0);

@override
Widget build(BuildContext context) {
   return Scaffold(
     appBar: AppBar(
       title: const Text('Syncfusion Flutter Range Slider'),
     ),
     body: SfRangeSlider(
        min: 0.0,
        max: 100.0,
        values: _values,
        interval: 20,
        showTicks: true,
        showLabels: true,
        enableTooltip: true,
        minorTicksPerInterval: 1,
        onChanged: (SfRangeValues values){
          setState(() {
            _values = values;
          });
        },
      ),
   );
}
```

The following screenshot illustrates the result of the above code sample.

![simple range slider](https://cdn.syncfusion.com/content/images/FTControl/Flutter/sliders/range-slider.png)

## Range selector getting started

Import the following package.

```dart
import 'package:syncfusion_flutter_sliders/sliders.dart';
```
### Add range selector to the widget tree

Add the range selector widget as a child of any widget. Here, the range selector widget is added as a child of the center widget.

```dart
@override
Widget build(BuildContext context) {
    return MaterialApp(
        home: Scaffold(
            body: Center(
                child: SfRangeSelector(),
            ),
        ),
    );
}
```

### Add range selector elements

Add a child of any type inside the range selector. Here, the Charts widget is added as a child of the range selector.

```dart
final DateTime dateMin = DateTime(2003, 01, 01);
final DateTime dateMax = DateTime(2010, 01, 01);
final SfRangeValues dateValues = SfRangeValues(DateTime(2005, 01, 01), DateTime(2008, 01, 01));

@override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Syncfusion Flutter Range Selector'),
      ),
      body: Container(
          margin: const EdgeInsets.all(0),
          padding: const EdgeInsets.all(0),
          child: Stack(
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.only(top: 10),
                child: Center(
                  child: SfRangeSelector(
                    min: dateMin,
                    max: dateMax,
                    initialValues: dateValues,
                    labelPlacement: LabelPlacement.betweenTicks,
                    interval: 1,
                    dateIntervalType: DateIntervalType.years,
                    dateFormat: DateFormat.y(),
                    showTicks: true,
                    showLabels: true,
                    child: Container(
                      child: SfCartesianChart(
                        margin: const EdgeInsets.all(0),
                        primaryXAxis: DateTimeAxis(
                          minimum: dateMin,
                          maximum: dateMax,
                          isVisible: false,
                        ),
                        primaryYAxis: NumericAxis(isVisible: false, maximum: 4),
                        series: <SplineAreaSeries<Data, DateTime>>[
                          SplineAreaSeries<Data, DateTime>(
                              dataSource: chartData,
                              xValueMapper: (Data sales, _) => sales.x,
                              yValueMapper: (Data sales, _) => sales.y)
                        ],
                      ),
                      height: 200,
                    ),
                  ),
                ),
              ),
            ],
          )),
    );
  }
```

The following screenshot illustrates the result of the above code sample.

![simple range selector](https://cdn.syncfusion.com/content/images/FTControl/Flutter/sliders/range-selector-image.png)

## Support and Feedback

* For any other queries, reach our [Syncfusion support team](https://www.syncfusion.com/support/directtrac/incidents/newincident) or post the queries through the [Community forums](https://www.syncfusion.com/forums) and submit a feature request or a bug through our [Feedback portal](https://www.syncfusion.com/feedback/flutter).
* To renew the subscription, click [renew](https://www.syncfusion.com/sales/products) or contact our sales team at salessupport@syncfusion.com | Toll Free: 1-888-9 DOTNET.

## About Syncfusion

Founded in 2001 and headquartered in Research Triangle Park, N.C., Syncfusion has more than 20,000 customers and more than 1 million users, including large financial institutions, Fortune 500 companies, and global IT consultancies.

Today we provide 1,000+ controls and frameworks for web ([ASP.NET Core](https://www.syncfusion.com/aspnet-core-ui-controls), [ASP.NET MVC](https://www.syncfusion.com/aspnet-mvc-ui-controls), [ASP.NET WebForms](https://www.syncfusion.com/jquery/aspnet-web-forms-ui-controls), [JavaScript](https://www.syncfusion.com/javascript-ui-controls), [Angular](https://www.syncfusion.com/angular-ui-components), [React](https://www.syncfusion.com/react-ui-components), [Vue](https://www.syncfusion.com/vue-ui-components), and [Blazor](https://www.syncfusion.com/blazor-components)), mobile ([Xamarin](https://www.syncfusion.com/xamarin-ui-controls), [Flutter](https://www.syncfusion.com/flutter-widgets), [UWP](https://www.syncfusion.com/uwp-ui-controls), and [JavaScript](https://www.syncfusion.com/javascript-ui-controls)), and desktop development ([WinForms](https://www.syncfusion.com/winforms-ui-controls), [WPF](https://www.syncfusion.com/wpf-ui-controls), and [UWP](https://www.syncfusion.com/uwp-ui-controls)). We provide ready-to deploy enterprise software for dashboards, reports, data integration, and big data processing. Many customers have saved millions in licensing fees by deploying our software.


CHANGELOG.md
## [18.4.30-beta] - 12/17/2020

## Slider

### Breaking changes

* The `showTooltip` property has been changed into `enableTooltip` property.

## Range Slider

### Breaking changes

* The `showTooltip` property has been changed into `enableTooltip` property.

## Range Selector

### Breaking changes

* The `showTooltip` property has been changed into `enableTooltip` property.

## [18.3.35-beta] - 10/01/2020

## Slider

### Features

* **Material design improvements** - The following UI improvements have been made based on the latest Material design guidelines: 

    * Difference in active and inactive track height.
    * Shadow support for the thumbs.
    * Changes in default divider radius.

## Range Slider

### Features

* **Material design improvements** - The following UI improvements have been made based on the latest Material design guidelines: 

    * Difference in default active and inactive track height.
    * Shadow support for the thumbs.
    * Overlapping stroke for thumb and tooltip.
    * Changes in default divider radius.

## Range Selector

### Features

* **Material design improvements** - The following UI improvements have been made based on the latest Material design guidelines: 

    * Difference in default active and inactive track height.
    * Shadow support for the thumbs.
    * Overlapping stroke for thumb and tooltip.
    * Changes in default divider radius.

## [18.2.59-beta] - 09/23/2020

No changes.

## [18.2.57-beta] - 09/08/2020

No changes.

## [18.2.56-beta] - 09/01/2020

No changes.

## [18.2.55-beta] - 08/25/2020

No changes.

## [18.2.54-beta] - 08/18/2020

No changes.

## [18.2.48-beta] - 08/04/2020

No changes.

## [18.2.47-beta] - 07/28/2020

No changes.

## [18.2.46-beta] - 07/21/2020

No changes.

## [18.2.45-beta] - 07/14/2020

No changes.

## [18.2.44-beta] - 07/07/2020

### Slider `Preview`

The Flutter Slider is a lightweight widget that allows you to select a single value from a range of values.

#### Features

* **Numeric and date support** - Provides functionality for selecting numeric and date values. For date, support is provided up to the seconds interval.
* **Labels** - Render labels for date and numeric values with the option to customize their formats based on your requirements.
* **Ticks and divisors** - Provides the option to show ticks and divisors based on the slider interval. Also, minor ticks can be enabled to indicate the values between each interval. These options present the selected value in a more intuitive way for end users.
* **Thumb icon support** - Accepts custom widgets like icon or text inside the thumb.
* **Tooltip** - Render a tooltip to show the selected value clearly. You can also customize the format of the text shown in the tooltip. Paddle and rectangular shape tooltips are supported.
* **Highly customizable** - In addition to the rich set of built-in features, the control is fully customizable through its wide range options.

### Range Slider

#### Features

* **Discrete selection** - Provides an option for selecting only discrete numeric and date values.
* **Interval selection** - Allows users to select a particular interval by tapping or clicking in it. Both thumbs will be moved to the current interval with smooth animation.
* **Thumb icon support** - Accepts custom widgets like icon or text inside the left and right thumbs.
* **Paddle tooltip** - Paddle shape tooltip support has been provided.
* Support has been provided to customize the radius of the active and inactive divisor using the `activeDivisorRadius` and `inactiveDivisorRadius` properties respectively in the `SfRangeSliderThemeData`.
* Support has been provided to customize the stroke width of the active and inactive divisor using the `activeDivisorStrokeWidth` and `inactiveDivisorStrokeWidth` properties respectively in the `SfRangeSliderThemeData`.
* Support has been provided to customize the stroke color of the active and inactive divisor using the `activeDivisorStrokeColor` and `inactiveDivisorStrokeColor` properties respectively in the `SfRangeSliderThemeData`.
* Support has been provided to customize the thumb stroke using the `thumbStrokeWidth` property in the `SfRangeSliderThemeData`.
* Support has been provided to customize the stroke color of the thumb using the `thumbStrokeColor` property in the `SfRangeSliderThemeData`.

#### Breaking changes

* The `trackHeight` property has been split into the `activeTrackHeight` and `inactiveTrackHeight` properties in the `SfRangeSliderThemeData`.

* The following arguments are added, removed and modified in the `getPreferredRect` method of `SfTrackShape`:

    * Modified the type of the argument parentBox from `RenderProxyBox` to `RenderBox`.
    * Added a new argument named `isActive`, which is used to denote whether currently, the active or inactive track is being drawn.
    * Removed the argument `isEnabled`.

* The following arguments are added, removed and modified in the `paint` method of `SfTrackShape`:

    * Modified the argument `animation` to `enableAnimation`.
    * Modified the type of the argument parentBox from `RenderProxyBox` to `RenderBox`.
    * Added a new argument named as `currentValues` that holds the current thumb values for `SfRangeSlider`.
    * Added a new argument named as `currentValue` that holds the current thumb value for `SfSlider`.
    * Added new arguments named as `activePaint` and `inactivePaint` for customizing the track.
    * Added a new argument named `thumbCenter` that holds the current thumb pixel position of `SfSlider`.
    * Removed the argument `isEnabled`.

* The following arguments are added, removed and modified in the `getPreferredRect` method of `SfThumbShape`:

    * Removed the argument `isEnabled`.

* The following arguments are added, removed and modified in the `paint` method of `SfThumbShape`:


common.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

/// Signature for formatting or changing the whole numeric or date label text.
typedef LabelFormatterCallback = String Function(

    /// actualValue will be either [DateTime] or [double]
    /// based on given [values].
    dynamic actualValue,

    /// If the actual value is [double], it is formatted by [numberFormat] and
    /// if the actual value is [DateTime], it is formatted by [dateFormat].
    String formattedText);

/// Signature for formatting or changing the whole tooltip label text.
typedef TooltipTextFormatterCallback = String Function(

    /// actualValue will be either [DateTime] or [double]
    /// based on given [values].
    dynamic actualValue,

    /// If the actual value is [double], it is formatted by [numberFormat] and
    /// if the actual value is [DateTime], it is formatted by [dateFormat].
    String formattedText);

/// The value will be either [double] or [DateTime] based on the `values`.
typedef SfSliderSemanticFormatterCallback = String Function(dynamic value);

typedef SfRangeSliderSemanticFormatterCallback = String Function(
    SfRangeValues values);

typedef SfRangeSelectorSemanticFormatterCallback = String Function(
    SfRangeValues values);

/// Option to place the labels either between the major ticks
/// or on the major ticks.
enum LabelPlacement {
  /// onTicks places the labels on the major ticks.
  onTicks,

  /// betweenTicks places the labels between the major ticks.
  betweenTicks
}

/// The type of date interval. It can be years to seconds.
enum DateIntervalType {
  /// Date interval is year.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 00) and
  /// `max` is DateTime(2005, 12, 31, 24) and `interval` is 1 and
  /// `dateIntervalType` is [years] then range slider
  /// will render labels for 2000, 2001, 2002, 2003, 2004, 2005 respectively.
  years,

  /// Date interval is month.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 00) and
  /// `max` is DateTime(2000, 12, 31, 24) and `interval` is 3 and
  /// `dateIntervalType` is [months] then range slider will render labels
  /// for [Jan 01, 2000], [Apr 01, 2000], [Jul 01, 2000], [Oct 01, 2000]
  /// and [Jan 01, 2001] respectively.
  months,

  /// Date interval is day.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 00) and
  /// `max` is DateTime(2000, 01, 25, 24) and `interval` is 5 and
  /// `dateIntervalType` is [days] then range slider will render labels
  /// for [Jan 01, 2000], [Jan 06, 2000], [Jan 11, 2000], [Jan 16, 2000],
  /// [Jan 21, 2001] and [Jan 26, 2001] respectively.
  days,

  /// Date interval is hour.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 09) and
  /// `max` is DateTime(2000, 01, 01, 17) and `interval` is 4 and
  /// `dateIntervalType` is [hours] then range slider will render labels for
  /// [Jan 01, 2000 09:00], [Jan 01, 2000 13:00], and [Jan 01, 2000 17:00]
  /// respectively.
  hours,

  /// Date interval is minute.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 09) and
  /// `max` is DateTime(2000, 01, 01, 10) and `interval` is 15 and
  /// `dateIntervalType` is [minutes] then range slider will render labels for
  /// [Jan 01, 2000 09:00], [Jan 01, 2000 09:15], [Jan 01, 2000 09:30],
  /// [Jan 01, 2000 09:45]and [Jan 01, 2000 10:00] respectively.
  minutes,

  /// Date interval is second.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 09, 00) and
  /// `max` is DateTime(2000, 01, 01, 09, 01) and `interval` is 20 and
  /// `dateIntervalType` is [seconds] then range slider will render labels for
  /// [Jan 01, 2000 09:00:00], [Jan 01, 2000 09:00:20], [Jan 01, 2000 09:00:40],
  /// and [Jan 01, 2000 09:01:00] respectively.
  seconds
}

/// Represents the [SfRangeSlider] or [SfRangeSelector] thumbs.
enum SfThumb {
  /// start represents the [SfRangeValues.start] thumb.
  start,

  /// end represents the [SfRangeValues.end] thumb.
  end
}

/// Represents the dragging behavior of the [SfRangeSelector] thumbs.
enum SliderDragMode {
  /// When [SliderDragMode] is set to [SliderDragMode.onThumb],
  /// individual thumb can be moved by dragging it.
  onThumb,

  /// When [SliderDragMode] is set to [SliderDragMode.betweenThumbs],
  /// both the thumbs can be moved at the same time by dragging in the area
  /// between start and end thumbs. The range between the start and end thumb
  /// will always be the same. Hence, it is not possible to move the
  /// individual thumb.
  betweenThumbs,

  /// When [SliderDragMode] is set to [SliderDragMode.both], individual thumb
  /// can be moved by dragging it, and also both the thumbs can be moved
  /// at the same time by dragging in the area between start and end thumbs.
  both
}

/// Represents the current selected values of [SfRangeSlider]
/// and [SfRangeSelector].
@immutable
class SfRangeValues extends DiagnosticableTree {
  /// Represents the current selected values of [SfRangeSlider]
  /// and [SfRangeSelector].
  const SfRangeValues(this.start, this.end);

  /// Represents the [SfRangeValues.start] thumb.
  final dynamic start;

  /// Represents the [SfRangeValues.end] thumb.
  final dynamic end;

  SfRangeValues copyWith({dynamic start, dynamic end}) {
    // HACK: In web, 0.00 and 0 are considered as identical.
    // So, we had considered both double and int.
    if (start != null &&
        (start.runtimeType == double || start.runtimeType == int) &&
        this.end.runtimeType == DateTime) {
      final double value = start;
      return SfRangeValues(
          DateTime.fromMillisecondsSinceEpoch(value.toInt()), end ?? this.end);
    } else if (end != null &&
        (end.runtimeType == double || end.runtimeType == int) &&
        this.start.runtimeType == DateTime) {
      final double value = end;
      return SfRangeValues(start ?? this.start,
          DateTime.fromMillisecondsSinceEpoch(value.toInt()));
    }

    return SfRangeValues(start ?? this.start, end ?? this.end);
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<dynamic>('start', start));
    properties.add(DiagnosticsProperty<dynamic>('end', end));
  }
}

/// The class which is used to set step duration for date discrete support.
///
/// See also:
///
/// * `interval`, for setting the interval.
class SliderStepDuration extends DiagnosticableTree {
  /// The discrete position is calculated by adding the arguments
  /// given in the [SliderStepDuration] object.
  /// By default, all arguments values are zero.
  const SliderStepDuration(
      {this.years = 0,
      this.months = 0,
      this.days = 0,
      this.hours = 0,
      this.minutes = 0,
      this.seconds = 0});

  /// Moves the thumbs based on years.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01) and
  /// `max` is DateTime(2006, 01, 01) and `SliderDuration(years: 2)` then the
  /// thumb will get moved at DateTime(2000, 01, 01), DateTime(2002, 07, 01),
  /// DateTime(2004, 01, 01), and DateTime(2006, 07, 01).
  final int years;

  /// Moves the thumbs based on months.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01) and
  /// `max` is DateTime(2000, 10, 01) and `SliderDuration(months: 3)` then the
  /// thumb will get moved at DateTime(2000, 01, 01), DateTime(2000, 04, 01),
  /// DateTime(2000, 07, 01), and DateTime(2000, 10, 01).
  final int months;

  /// Moves the thumbs based on days.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01) and
  /// `max` is DateTime(2000, 01, 20) and `SliderDuration(days: 5)` then the
  /// thumb will get moved at DateTime(2000, 01, 01), DateTime(2000, 01, 06),
  /// DateTime(2000, 01, 11), DateTime(2000, 01, 16), and
  /// DateTime(2000, 01, 20).
  final int days;

  /// Moves the thumbs based on hours.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 05) and
  /// `max` is DateTime(2006, 01, 01, 20) and `SliderDuration(hours: 5)` then
  /// the thumb will get moved at DateTime(2000, 01, 01, 05),
  /// DateTime(2000, 01, 01, 11), DateTime(2000, 01, 01, 16),
  /// and DateTime(2000, 01, 01, 20).
  final int hours;

  /// Moves the thumbs based on minutes.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 05, 10) and
  /// `max` is DateTime(2006, 01, 01, 05, 40) and `SliderDuration(minutes: 10)`
  /// then the thumb will get moved at DateTime(2000, 01, 01, 05, 10),
  /// DateTime(2000, 01, 01, 05, 20), DateTime(2000, 01, 01, 05, 30),
  /// and DateTime(2000, 01, 01, 05, 40).
  final int minutes;

  /// Moves the thumbs based on seconds.
  ///
  /// For example, if `min` is DateTime(2000, 01, 01, 05, 10, 00) and
  /// `max` is DateTime(2006, 01, 01, 05, 12, 00) and
  /// `SliderDuration(seconds: 30)` then the thumb will get moved at
  /// DateTime(2006, 01, 01, 05, 10, 00), DateTime(2006, 01, 01, 05, 10, 30)),
  /// DateTime(2006, 01, 01, 05, 11, 00), DateTime(2006, 01, 01, 05, 11, 30),
  /// and DateTime(2006, 01, 01, 05, 12, 00).
  final int seconds;

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);

    properties.add(IntProperty('years', years));
    properties.add(IntProperty('months', months));
    properties.add(IntProperty('days', days));
    properties.add(IntProperty('hours', hours));
    properties.add(IntProperty('minutes', minutes));
    properties.add(IntProperty('seconds', seconds));
  }
}
